File: branches/Orca/src/orc/lib/state/types/ArrayType.java
//
// ArrayType.java -- Java class ArrayType
// Project OrcJava
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state.types;

import java.util.List;

import orc.error.compiletime.typing.TypeException;
import orc.type.Type;
import orc.type.structured.ArrowType;
import orc.type.structured.DotType;
import orc.type.tycon.MutableContainerType;

public class ArrayType extends MutableContainerType {

	@Override
	public String toString() {
		return "Array";
	}

	@Override
	public Type makeCallableInstance(final List<Type> params) throws TypeException {
		final Type T = params.get(0);

		/* Default behavior is element reference retrieval */
		final Type RefOfT = new RefType().instance(T);
		final DotType arrayType = new DotType(new ArrowType(Type.INTEGER, RefOfT));

		final Type ArrayOfT = new ArrayType().instance(T);
		arrayType.addField("get", new ArrowType(Type.INTEGER, T));
		arrayType.addField("set", new ArrowType(Type.INTEGER, T, Type.SIGNAL));
		arrayType.addField("slice", new ArrowType(Type.INTEGER, Type.INTEGER, ArrayOfT));
		arrayType.addField("length", new ArrowType(Type.INTEGER));
		arrayType.addField("fill", new ArrowType(T, Type.SIGNAL));
		return arrayType;
	}

}

File: OrcScala/src/orc/lib/state/types/ArrayType.scala
//
// ArrayType.scala -- Scala object ArrayType
// Project OrcScala
//
// Created by dkitchin on Dec 1, 2010.
//
// Copyright (c) 2011 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//
package orc.lib.state.types

import orc.types._
import orc.error.compiletime.typing._

/**
  * @author dkitchin
  */
object ArrayType extends SimpleTypeConstructor("Array", Invariant) {

  def getBuilder: Type = {

    val X = new TypeVariable()
    val makeEmpty = FunctionType(List(X), List(IntegerType), this(X))

    val Y = new TypeVariable()
    val makeFull = FunctionType(List(Y), List(IntegerType, StringType), this(Y))

    OverloadedType(List(makeEmpty, makeFull))
  }

  override def instance(ts: List[Type]) = {
    val List(t) = ts
    new RecordType(
      "apply" -> SimpleFunctionType(IntegerType, RefType(t)),
      "length" -> RefType(IntegerType))
  }

}
